src/Compilers/Core/Portable/SourceGeneration/Nodes/IncrementalValueSources.cs

```
@@ -0,0 +1,75 @@
+
+        //only used for back compat in the adaptor
+        internal IncrementalValueSource<ISyntaxContextReceiver> CreateSyntaxReceiver() => new IncrementalValueSource<ISyntaxContextReceiver>(_perGeneratorBuilder.GetOrCreateReceiverNode());
+    }
+
+    /// <summary>
+    /// Holds input nodes that are shared between generators and always exist
+    /// </summary>
+    internal static class SharedInputNodes
+    {
+        public static readonly InputNode<Compilation> Compilation = new InputNode<Compilation>();
+
+        public static readonly InputNode<ParseOptions> ParseOptions = new InputNode<ParseOptions>();
+
+        public static readonly InputNode<AdditionalText> AdditionalTexts = new InputNode<AdditionalText>();
+
+        public static readonly InputNode<AnalyzerConfigOptionsProvider> AnalzerConfigOptions = new InputNode<AnalyzerConfigOptionsProvider>();
+    }
+
+    /// <summary>
+    /// Holds input nodes that are created per-generator
+    /// </summary>
+    internal sealed class PerGeneratorInputNodes
+    {
+        public static PerGeneratorInputNodes Empty = new PerGeneratorInputNodes();
+
+        private PerGeneratorInputNodes() { }
+
+        private PerGeneratorInputNodes(InputNode<ISyntaxContextReceiver>? receiverNode)
+        {
+            this.ReceiverNode = receiverNode;
+        }
+
+        public InputNode<ISyntaxContextReceiver>? ReceiverNode { get; }
+
+        public sealed class Builder
+        {
+            private InputNode<ISyntaxContextReceiver>? _receiverNode;
+
+            public Builder()
+            {
+            }
+
+            public InputNode<ISyntaxContextReceiver> GetOrCreateReceiverNode() => InterlockedOperations.Initialize(ref _receiverNode, new InputNode<ISyntaxContextReceiver>());
+
+            public PerGeneratorInputNodes ToImmutable() => new PerGeneratorInputNodes(_receiverNode);
```

As constructed this builder could be used after the `ToImmutable()` call and the code would fail silently. It took me a bit of work to convince myself the code was correct here. Think it would be worth reworking the builder so that `ToImmutable` was destructive such that future calls to `GetOrCreateReceiverNode` failed 

Also shouldn't this return `PerGeneratorInptuNodes.Empty` when `_receiverNode is null` to avoid unnecessary allocations?